### PR TITLE
[Loh Wei-Ming, Darin] ip

### PR DESCRIPTION
**Pull request by talkintomato**
# Duke is the task manager of task managers 
> The task manager your task manager uses :sunglasses:

## Reasons to use Duke: 
1. It will _never_ give you up 
2. It will _never_ let you down 

## Things Duke cannot do: 
- Feed your cat 
- [Look at all those chickens ](https://www.youtube.com/watch?v=NsLKQTh-Bqo&ab_channel=rawuncutvines)

## Features: 
- [x] create tasks 
- [ ] set deadlines 
- [ ] be your friend 

``` java
public class Duke {
    Run();
}
```

### CODE REVIEW

File: src/main/java/duke/Command.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import duke.tasks.*;
+
+import java.time.format.DateTimeParseException;
+
+
+public class Command {
+
+    private String[] cmd;
+
+    private boolean isExit = false;
```
Comment (KeithCZW) [inline src/main/java/duke/Command.java]: I feel that isExit is not descriptive enough. Is the command object an exit of what? Perhaps isExiting would be a better name for this.

File: src/main/java/duke/Command.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                createEvent(cmd, taskList, ui);
+                break;
+            case "delete":
+                taskList.removeTask(cmd[1]);
+                break;
+            case "bye" :
+                isExit = true;
+                break;
+            default:
+                throw new DukeException("I'm not sure what that means :((");
```
Comment (KeithCZW) [inline src/main/java/duke/Command.java]: The indentations are off for your switch cases. The case should be inline with the switch function.

File: src/main/java/duke/Parser.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package duke;
+
+public class Parser {
+
+
```
Comment (KeithCZW) [inline src/main/java/duke/Parser.java]: Just a small nitpick, but there's an additional space here. 😅

File: src/main/java/duke/Storage.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import duke.exception.DukeException;
+import duke.tasks.*;
+
+public class Storage {
+
+    protected ArrayList<Task> todos = new ArrayList<>();
+    private String filepath;
```
Comment (KeithCZW) [inline src/main/java/duke/Storage.java]: filepath should be filePath to align with camel case.

File: src/main/java/duke/Storage.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                    Todo currTodo = (Todo) currTask;
+                    myWriter.write(currTodo.getInitial() + " , " + currTodo.getIsDone() + " , " + currTodo.getDescription() + " , " + currTodo.getTime() +"\n");
+                }
+                else if (currTask instanceof Event) {
+                    Event currEvent = (Event) currTask;
+                    myWriter.write(currEvent.getInitial() + " , " + currEvent.getIsDone() + " , " + currEvent.getDescription() + " , " + currEvent.getTime() +"\n");
+                }
+                else if (currTask instanceof Deadline) {
+                    Deadline currDeadline = (Deadline) currTask;
+                    myWriter.write(currDeadline.getInitial() + " , " + currDeadline.getIsDone() + " , " + currDeadline.getDescription() + " , " + currDeadline.getTime() +"\n");
```
Comment (KeithCZW) [inline src/main/java/duke/Storage.java]: This line is more than 120 characters and should be split into two lines.

**Review by KeithCZW — COMMENTED** [review]
Just a few minor issues with the PR but overall it's looking good! Just a few small parts to make changes to in order to follow the coding standard. Good job!

File: src/main/java/duke/Command.java (hunk 2 above)
Comment (talkintomato) [reply src/main/java/duke/Command.java]: @KeithCZW Thanks for the feedback! do u have a link which shows the case and switch being inline? most of the examples I have googled have shown otherwise.

Comment (KeithCZW) [reply src/main/java/duke/Command.java]: https://se-education.org/guides/conventions/java/intermediate.html#layout

Under layout section, switch and cases are aligned.

Comment (talkintomato) [reply src/main/java/duke/Command.java]: Thanks!

File: src/main/java/duke/Command.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
+package duke;
+import duke.exception.DukeException;
+import duke.tasks.*;
```
Comment (jaysmyname) [inline src/main/java/duke/Command.java]: Shouldn't imported classes be listed explicitly?

File: src/main/java/duke/Command.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        case "bye" :
+            isExiting = true;
+            break;
+        default:
+            throw new DukeException("I'm not sure what that means :((");
+        }
+        storage.saveList();
+    }
+
+    public boolean getExitStatus() {
```
Comment (jaysmyname) [inline src/main/java/duke/Command.java]: Perhaps change the boolean method to sound more like a boolean? Something like shouldExit()

File: src/main/java/duke/Storage.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                }
+                else if (currTask instanceof Event) {
+                    Event currEvent = (Event) currTask;
+                    myWriter.write(currEvent.getInitial() + " , " + currEvent.getIsDone() + " , "
+                            + currEvent.getDescription() + " , " + currEvent.getTime() +"\n");
+                }
+                else if (currTask instanceof Deadline) {
+                    Deadline currDeadline = (Deadline) currTask;
+                    myWriter.write(currDeadline.getInitial() + " , " + currDeadline.getIsDone() + " , "
+                            + currDeadline.getDescription() + " , " + currDeadline.getTime() +"\n");
```
Comment (jaysmyname) [inline src/main/java/duke/Storage.java:92]: Perhaps add a method to each task type to return a string that has all the gets so you can simply do one get call?

File: src/main/java/duke/Command.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            break;
+        case "delete":
+            taskList.removeTask(cmd[1]);
+            break;
+        case "bye" :
+            isExiting = true;
+            break;
+        default:
+            throw new DukeException("I'm not sure what that means :((");
+        }
```
Comment (jaysmyname) [inline src/main/java/duke/Command.java]: I like how concise this is! Very compact.

File: src/main/java/duke/Storage.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+import duke.exception.DukeException;
+import duke.tasks.*;
```
Comment (jaysmyname) [inline src/main/java/duke/Storage.java:11]: Shouldn't imported classes be listed explicitly?

**Review by jaysmyname — COMMENTED** [review]
Overall really good job following the coding standards!

File: src/main/java/duke/Command.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+
+    public boolean getExitStatus() {
+        return this.isExiting;
+    }
+
+//    public void validateTaskCreation(String[] cmd) {
+//        if (cmd.length != 2) {
+//    }
```
Comment (Oddcorner) [inline src/main/java/duke/Command.java]: Perhaps it would be better to remove this code entirely instead of leaving it commented. I noticed a few other places where quite a large number of unused code lines are commented too.

File: src/main/java/duke/Command.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                } catch (DateTimeParseException e) {
+                    System.out.println(e.getMessage() + "\nPlease input deadline in yyyy-mm-dd");
+                }
+            } else {
+                throw new DukeException("Please indicate a time for the Deadline!");
+            }
+        } else {
+            throw new DukeException("The description of a Deadline date cannot be empty.");
+        }
+        return response;
```
Comment (Oddcorner) [inline src/main/java/duke/Command.java]: Perhaps this nested if statements could be refactored with guard clauses to make the happy path more apparent. Similarily with createEvent().
```suggestion
        String response = "";
        if (cmd.length < 2) {
            throw new DukeException("The description of a Deadline date cannot be empty.");
        }

        String[] deadlineDetails = cmd[1].split(" /by ", 2);

        if (deadlineDetails.length < 2) {
                throw new DukeException("Please indicate a time for the Deadline!");
        }

        try {
            Deadline deadline = new Deadline(deadlineDetails[0], deadlineDetails[1]);
            taskList.addTask(deadline);
            response = ui.showTaskAdded(deadline);
        } catch (DateTimeParseException e) {
            System.out.println(e.getMessage() + "\nPlease input deadline in yyyy-mm-dd");
        }
   
        return response;
```

File: src/main/java/duke/tasks/Task.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            case "D":
+                return new Deadline(description[1], description[2], description[3]);
+            default:
+                throw new DukeException("Invalid Task");
+        }
+    }
+
+    abstract public String getTime();
+
+    public Boolean getIsDone() {return this.isDone;}
```
Comment (Oddcorner) [inline src/main/java/duke/tasks/Task.java:50]: I think naming this function as "isDone" is more in line with the coding quality naming convention, what do you think?

File: src/main/java/duke/tasks/Task.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        switch (description[0]) {
+            case "T":
+                return new Todo(description[1], description[2]);
+            case "E":
+                return new Event(description[1], description[2], description[3]);
+            case "D":
+                return new Deadline(description[1], description[2], description[3]);
+            default:
+                throw new DukeException("Invalid Task");
+        }
```
Comment (Oddcorner) [inline src/main/java/duke/tasks/Task.java:45]: switch and case statements should have the same indentation level to follow the coding standard guidelines.

**Review by Oddcorner — COMMENTED** [review]
Overall code quality looks good! 👍 
The only places I found hard to follow/read is how the file is read/written in Stroage.java
Perhaps some method extractions will help with code quality there!

Good Job!
